homeassistant/components/integration/sensor.py

```diff
@@ -291,109 +411,101 @@ async def async_added_to_hass(self) -> None:
             self._attr_device_class = state.attributes.get(ATTR_DEVICE_CLASS)
             self._unit_of_measurement = state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
 
-        @callback
-        def calc_integration(event: EventType[EventStateChangedData]) -> None:
-            """Handle the sensor state changes."""
-            old_state = event.data["old_state"]
-            new_state = event.data["new_state"]
-
-            if (
-                source_state := self.hass.states.get(self._sensor_source_id)
-            ) is None or source_state.state == STATE_UNAVAILABLE:
-                self._attr_available = False
-                self.async_write_ha_state()
-                return
-
-            self._attr_available = True
-
-            if old_state is None or new_state is None:
-                # we can't calculate the elapsed time, so we can't calculate the integral
-                return
-
-            unit = new_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
-            if unit is not None:
-                self._unit_of_measurement = self._unit(unit)
-
-            if (
-                self.device_class is None
-                and new_state.attributes.get(ATTR_DEVICE_CLASS)
-                == SensorDeviceClass.POWER
-            ):
-                self._attr_device_class = SensorDeviceClass.ENERGY
-                self._attr_icon = None
+        if self._max_dt is not None:
+            self._schedule_max_dt_exceeded()
+            self.async_on_remove(self._cancel_max_dt_exceeded_callback)
 
+        self.async_on_remove(
+            async_track_state_change_event(
+                self.hass,
+                [self._sensor_source_id],
+                self._integrate_on_state_change_and_max_dt
+                if self._max_dt is not None
+                else self._integrate_on_state_change,
+            )
+        )
+
+    @callback
+    def _integrate_on_state_change_and_max_dt(
+        self, event: EventType[EventStateChangedData]
+    ) -> None:
+        try:
+            self._cancel_max_dt_exceeded_callback()
+            self._integrate_on_state_change(event)
+            self._last_integration_trigger = _IntegrationTrigger.StateChange
+            self._last_integration_time = datetime.now(tz=UTC)
+        finally:
+            self._schedule_max_dt_exceeded()
```

Comment:
Should we skip calling this if the state is `0`? I use riemann a lot for sensors that spend a lot of time at 0, and it would be nice to have high frequency updates when the sensor is nonzero, but seems just wasted cycles to keep calling this when it is not accumulating.

Reply:
Sounds like a nice optimization! Will give it a try.

Reply:
Don't over optimize... this integration is doing simple math, it takes no time. 

Go for it, but don't overcomplicate the code ;)

Reply:
While the individual operations are probably minor simple math, I wasn't sure if having dozens of riemann's doing updates frequently forever would put extra load on the system or reduce power optimization if it has to keep waking up all the time to service and reschedule timers. 

But I'm not a core expert, so feel free to ignore the suggestion if it doesn't make sense. 

Just thinking about my own use case, I have 20 sprinkler zones which are off/`0` 99% of the time, but use riemann to accumulate water consumption when they are on. It would be cool if they could update consumption in relatively quick realtime when they are on (once/minute?), but didn't want to have them all calling callback timers infinitely forever.

Reply:
You should not worry about it :) specially because max_dt will work seconds (not milliseconds)

Reply:
Yes keep it simple. Now I use helper template sensors with time triggers to force updates. This will be more efficient I guess.
Thanks for fixing this 🙏

Reply:
On second thought this cannot be skipped when the source is 0. We still want to update `_last_integration_time` and `_last_integration_trigger` even if the source is constant 0. This doesn't affect the integral, but it will affect future calculations if your source starts increasing again after a period of constant 0. 

Did apply the following other minor optimizations:
- Using the new state for scheduling a trigger and not accessing the source sensor from hass. Because if the timer didn't cancel, the state wasn't changed, so the earlier obtained new_state from the `EventStateChangedData` can be used. This is more clean and in line with the source being constant assumption.
- Only schedule a trigger if the new_state is numeric.